src/flint/fmpz.jl

```diff
@@ -902,27 +903,65 @@ function sqrtmod(x::fmpz, m::fmpz)
     return z
 end
 
+function _normalize_crt(r::fmpz, m::fmpz, signed)
+   s = sign(Int, m)
+   if s > 0
+      return signed ? nfdivrem(r, m)[2] : fdivrem(r, m)[2]
+   elseif s < 0
+      return signed ? ncdivrem(r, m)[2] : cdivrem(r, m)[2]
+   else
+      return r
+   end
+end
+
+function _normalize_crt_with_lcm(r::fmpz, m::fmpz, signed)
+   s = sign(Int, m)
+   if s == 0
+      return (r, m)
+   elseif s < 0
+      m = -m
+   end
+   return (signed ? nfdivrem(r, m)[2] : fdivrem(r, m)[2], m)
+end
+
 @doc Markdown.doc"""
-    crt(r1::fmpz, m1::fmpz, r2::fmpz, m2::fmpz, signed=false)
+    crt(r1::fmpz, m1::fmpz, r2::fmpz, m2::fmpz, signed=false; check::Bool=true)
+    crt(r::Vector{fmpz}, m::Vector{fmpz}, signed=false; check::Bool=true)
+    crt_with_lcm(r1::fmpz, m1::fmpz, r2::fmpz, m2::fmpz, signed=false; check::Bool=true)
+    crt_with_lcm(r::Vector{fmpz}, m::Vector{fmpz}, signed=false; check::Bool=true)
 
-Return $r$ such that $r \equiv r_1 (\mod m_1)$ and $r \equiv r_2 (\mod m_2)$.
-If `signed = true`, $r$ will be in the range $-m_1m_2/2 < r \leq m_1m_2/2$.
-If `signed = false` the value will be in the range $0 \leq r < m_1m_2$.
+As per the AbstractAlgebra `crt` interface, with the following option.
+If `signed = true`, the solution is the range (-m/2, m/2], otherwise it is in
+the range `[0,m)`, where `m` is the least common multiple of the moduli.
 """
-function crt(r1::fmpz, m1::fmpz, r2::fmpz, m2::fmpz, signed=false)
-   z = fmpz()
-   ccall((:fmpz_CRT, libflint), Nothing,
-          (Ref{fmpz}, Ref{fmpz}, Ref{fmpz}, Ref{fmpz}, Ref{fmpz}, Cint),
-          z, r1, m1, r2, m2, signed)
-   return z
+function crt(r1::fmpz, m1::fmpz, r2::fmpz, m2::fmpz, signed=false; check::Bool=true)
+   r, m = AbstractAlgebra._crt_with_lcm_stub(r1, m1, r2, m2; check=check)
+   return _normalize_crt(r, m, signed)
+end
+
+function crt(r::Vector{fmpz}, m::Vector{fmpz}, signed=false; check::Bool=true)
+   r, m = AbstractAlgebra._crt_with_lcm_stub(r, m; check=check)
+   return _normalize_crt(r, m, signed)
 end
 
+function crt_with_lcm(r1::fmpz, m1::fmpz, r2::fmpz, m2::fmpz, signed=false; check::Bool=true)
+   r, m = AbstractAlgebra._crt_with_lcm_stub(r1, m1, r2, m2; check=check)
+   return _normalize_crt_with_lcm(r, m, signed)
+end
+
+function crt_with_lcm(r::Vector{fmpz}, m::Vector{fmpz}, signed=false; check::Bool=true)
+   r, m = AbstractAlgebra._crt_with_lcm_stub(r, m; check=check)
+   return _normalize_crt_with_lcm(r, m, signed)
+end
+
+
 @doc Markdown.doc"""
     crt(r1::fmpz, m1::fmpz, r2::Int, m2::Int, signed=false)
 
 Return $r$ such that $r \equiv r_1 (\mod m_1)$ and $r \equiv r_2 (\mod m_2)$.
 If `signed = true`, $r$ will be in the range $-m_1m_2/2 < r \leq m_1m_2/2$.
 If `signed = false` the value will be in the range $0 \leq r < m_1m_2$.
+This function expects $m_1$ and $m_2$ to be coprime.
```

Comment:
no

Reply:
Ah, I see. This is just some more "low-level" version, useful maybe for iterative stuff? The version with all arguments of type `fmpz` does not have this restriction, right? 

I find it is just a bit surprising that
```julia
julia> crt(2, 4, 0, 2)
2
```
works, but
```julia
julia> crt(ZZ(2), ZZ(4), 0, 2)
```
might fail horribly without any error.

Reply:
hehe `crt(2, 4, 0, 2)` only works because of the bogus ring `zz`.
I think you would only use these functions for experiments because there is no precomputation, so I will give `crt(ZZ(2), ZZ(4), 0, 2)` a try.